File: Dictionaries and Sets/49. Group Anagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        anagrams = defaultdict(list)
        for s in strs:
            k = "".join(sorted(s))
            anagrams[k].append(s)
        return list(anagrams.values())

# Time: O(n * m log m)
# Space: O(n * m)

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        anagrams = defaultdict(list)
        for s in strs:
            alphabet = [0] * 26
            for c in s:
                alphabet[ord(c) - ord("a")] += 1
            anagrams[tuple(alphabet)].append(s)
        return list(anagrams.values())

# Time: O(n * m)
# Space: O(n * m)

File: 49. Group Anagrams/49. Group Anagrams.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {

  public List<List<String>> groupAnagrams(String[] strs) {
    Map<String, List<String>> anagrams = new HashMap<>();

    for (String s : strs) {
      int[] alphabet = new int[26];
      for (char c : s.toCharArray()) {
        alphabet[c - 'a']++;
      }
      StringBuilder key = new StringBuilder();
      for (int num : alphabet) {
        key.append(num).append(',');
      }
      String keyString = key.toString();
      anagrams.computeIfAbsent(keyString, k -> new ArrayList<>()).add(s);
    }

    return new ArrayList<>(anagrams.values());
  }
}
